### PR TITLE
[receiver/vcenter] Adds vCenter VM CPU readiness metric

### PR DESCRIPTION
**Pull request by BominRahmani**
**Description:** <Describe what has changed.>

This PR adds the `vcenter.vm.cpu.readiness` metric.
More information on this metric can be found [here](https://vdc-repo.vmware.com/vmwb-repository/dcr-public/d1902b0e-d479-46bf-8ac9-cee0e31e8ec0/07ce8dbd-db48-4261-9b8f-c6d3ad8ba472/vim.vm.Summary.QuickStats.html).
**Link to tracking Issue:** #33607

**Testing:** <Describe what testing was performed and which tests were added.>
The metric was scraped from a test vCenter environment, and golden test files were updated to reflect the addition of the metric.

**Documentation:** <Describe the documentation added.>
Documentation was updated according to the metadata.yaml

### CODE REVIEW

File: receiver/vcenterreceiver/metadata.yaml [hunk 1]
```diff
@@ -441,6 +441,13 @@ metrics:
       value_type: int
       aggregation_temporality: cumulative
     attributes: []
+  vcenter.vm.cpu.readiness:
+    enabled: true
```
Comment (StefanKurek) [inline receiver/vcenterreceiver/metadata.yaml:445]: I believe this is going to have to start as false for a new addition.

This will also need a warning that it will be enabled by default in two releases (https://github.com/open-telemetry/opentelemetry-collector/blob/main/cmd/mdatagen/metadata-schema.yaml)

File: receiver/vcenterreceiver/testdata/metrics/expected-all-enabled.yaml [hunk 2]
```diff
@@ -150,12 +150,9 @@ resourceMetrics:
         - key: vcenter.datacenter.name
           value:
             stringValue: Datacenter
-        - key: vcenter.cluster.name
```
Comment (StefanKurek) [inline receiver/vcenterreceiver/testdata/metrics/expected-all-enabled.yaml:153]: @BominRahmani were these changes necessary?

File: receiver/vcenterreceiver/testdata/metrics/expected.yaml [hunk 3]
```diff
@@ -150,12 +150,9 @@ resourceMetrics:
         - key: vcenter.datacenter.name
           value:
             stringValue: Datacenter
-        - key: vcenter.cluster.name
```
Comment (StefanKurek) [inline receiver/vcenterreceiver/testdata/metrics/expected.yaml:153]: Same, were all these changes necessary?

File: receiver/vcenterreceiver/testdata/metrics/expected.yaml [hunk 4]
```diff
@@ -6450,26 +6463,37 @@ resourceMetrics:
...
+            stringValue: CentOS 8
     scopeMetrics:
       - metrics:
+          - description: Percentage of time that the virtual machine was ready, but could not get scheduled to run on the physical CPU.
+            gauge:
+              dataPoints:
+                - asInt: "0"
+                  startTimeUnixNano: "1000000"
+                  timeUnixNano: "2000000"
+            name: vcenter.vm.cpu.readiness
```
Comment (StefanKurek) [inline receiver/vcenterreceiver/testdata/metrics/expected.yaml:6495]: Does this pass? That's cool if you didn't need to modify the mock api files.

File: receiver/vcenterreceiver/metrics.go [hunk 5]
```diff
@@ -140,6 +141,23 @@ func (v *vcenterMetricScraper) recordVMStats(
 		return
 	}
 	v.mb.RecordVcenterVMCPUUtilizationDataPoint(ts, 100*float64(cpuUsage)/float64(cpuLimit))
+
+	// OverallCpuReadiness is only available in vSphere API 7.0
+	// https://vdc-repo.vmware.com/vmwb-repository/dcr-public/d1902b0e-d479-46bf-8ac9-cee0e31e8ec0/07ce8dbd-db48-4261-9b8f-c6d3ad8ba472/vim.vm.Summary.QuickStats.html
+	vsphereAPIVersion, err := version.NewVersion(v.client.vimDriver.ServiceContent.About.ApiVersion)
```
Comment (StefanKurek) [inline receiver/vcenterreceiver/metrics.go:147]: is this making a request for every VM? might be overkill. Could you just check if cpuReadiness is null or something and not record it in that case?

**Review by StefanKurek — CHANGES_REQUESTED** [review]
@BominRahmani Overall looks OK. Have a few questions though. Also think it needs an integration test update.

File: receiver/vcenterreceiver/metrics.go [hunk 6]
```diff
@@ -140,6 +141,23 @@ func (v *vcenterMetricScraper) recordVMStats(
 		return
 	}
 	v.mb.RecordVcenterVMCPUUtilizationDataPoint(ts, 100*float64(cpuUsage)/float64(cpuLimit))
+
+	// OverallCpuReadiness is only available in vSphere API 7.0
+	// https://vdc-repo.vmware.com/vmwb-repository/dcr-public/d1902b0e-d479-46bf-8ac9-cee0e31e8ec0/07ce8dbd-db48-4261-9b8f-c6d3ad8ba472/vim.vm.Summary.QuickStats.html
+	vsphereAPIVersion, err := version.NewVersion(v.scrapeData.apiVersion)
```
Comment (StefanKurek) [inline receiver/vcenterreceiver/metrics.go:147]: I went back and forth on whether I was going to comment more on this or just leave it be.

It seems like this logic could be moved to a new function (or maybe even to the client as a helper function). The client could be reworked to have a private method like "refreshVsphereAPIVersion" which stores its own version. We could make sure we call this once a collection (or on client initialization). Then we could have another method `vsphereAPIVersionMeetsMin(minVsphereAPIVersion string) bool` that would be used here where you just pass in "7.0.0". It could do all this logic (making use of the stored version (possibly calling `refresh` if it's empty) and then return a true/false for the conditional for lines 158/159. This would be reusable for other cases too.

Ultimately I think I'm fine with it as is, just thinking that it feels a little out of place here and would likely be better placed elsewhere.

Comment (BominRahmani) [reply receiver/vcenterreceiver/metrics.go:147]: This is probably a good idea, there may be more of these metrics that require a check on the API in the future. I'll look into changing it up!

**Review by jlg-io — COMMENTED** [review]
Great work!

Maybe we could update `vcenterreceiver/README.md` to either drop support for `6.7` or add a comment that `overallCpuReadiness` won't work with this version?

File: receiver/vcenterreceiver/metrics.go (hunk 6 above)
Comment (StefanKurek) [reply receiver/vcenterreceiver/metrics.go:147]: @jlg-io That JSON API is a totally different (and would require major changes to the vcenterreceiver) from how we're currently collecting. In addition, it looks like it's only been available since vCenter 8.0+. I do agree that a note in the README would be a good add.

Comment (jlg-io) [reply receiver/vcenterreceiver/metrics.go:147]: @StefanKurek, my bad! I guess the right one is there: https://dp-downloads.broadcom.com/api-content/apis/API_VMA_001/8.0U2/html/vim.vm.Summary.QuickStats.html

Comment (BominRahmani) [reply receiver/vcenterreceiver/metrics.go:147]: Thanks! I'll add in a section in the readme and also update the link in the comments.

File: receiver/vcenterreceiver/README.md [hunk 7]
```diff
@@ -55,4 +55,7 @@ The full list of settings exposed for this receiver are documented [here](./conf
 
 Details about the metrics produced by this receiver can be found in [metadata.yaml](./metadata.yaml) with further documentation in [documentation.md](./documentation.md)
 
+Additionally some metrics are only available to certain vSphere API versions.
+- [vcenter.vm.cpu.readiness](https://dp-downloads.broadcom.com/api-content/apis/API_VMA_001/8.0U2/html/vim.vm.Summary.QuickStats.html)
```
Comment (StefanKurek) [inline receiver/vcenterreceiver/README.md:59]: Small thing, but I think having `7.0.0+` as a header for this list would be appropriate.

File: receiver/vcenterreceiver/metrics.go [hunk 8]
```diff
@@ -140,6 +141,27 @@ func (v *vcenterMetricScraper) recordVMStats(
...
+
+	// OverallCpuReadiness is only available in vSphere API 7.0
+	// https://dp-downloads.broadcom.com/api-content/apis/API_VMA_001/8.0U2/html/vim.vm.Summary.QuickStats.html
+	if v.vsphereAPIVersionMeetsMin("7.0.0") {
+		cpuReadiness := vm.Summary.QuickStats.OverallCpuReadiness
+		v.mb.RecordVcenterVMCPUReadinessDataPoint(ts, int64(cpuReadiness))
+	}
+}
+
+func (v *vcenterMetricScraper) vsphereAPIVersionMeetsMin(minVsphereAPIVersion string) bool {
```
Comment (StefanKurek) [inline receiver/vcenterreceiver/metrics.go:153]: Minor: it still feels like this function belongs in the client a little more than here. I don't feel strong enough to halt the PR based on this though, so I'll let you decide.

File: receiver/vcenterreceiver/metadata.yaml (hunk 1 above)
Comment (StefanKurek) [reply receiver/vcenterreceiver/metadata.yaml]: Can resolve

File: receiver/vcenterreceiver/testdata/metrics/expected.yaml (hunk 4 above)
Comment (StefanKurek) [reply receiver/vcenterreceiver/testdata/metrics/expected.yaml]: Can resolve

**Review by StefanKurek — CHANGES_REQUESTED** [review]

File: receiver/vcenterreceiver/client.go [hunk 9]
```diff
@@ -103,6 +106,24 @@ func (vc *vcenterClient) Datacenters(ctx context.Context) ([]mo.Datacenter, erro
 	return datacenters, nil
 }
 
+func (vc *vcenterClient) getVsphereAPIVersion() string {
+	return vc.vimDriver.ServiceContent.About.ApiVersion
+}
+
+func (vc *vcenterClient) vsphereAPIVersionMeetsMin(minVsphereAPIVersion string) bool {
```
Comment (StefanKurek) [inline receiver/vcenterreceiver/client.go:113]: @BominRahmani I like this better here! Should be made public and then could use a few additional unit tests (< > = versions). 😄  Sorry to keep this going. After that should be good though!

**Review by StefanKurek — APPROVED** [review]
Looks good enough to move out of Draft form. Thanks!

File: receiver/vcenterreceiver/README.md (hunk 7 above)
Comment (djaglowski) [reply receiver/vcenterreceiver/README.md:60]: I'm not sure how accurate [this page](https://endoflife.date/vcenter) is, but it claims that 6.7 is EOL, so perhaps we should just drop support?

File: receiver/vcenterreceiver/client.go [hunk 10]
```diff
@@ -74,6 +76,7 @@ func (vc *vcenterClient) EnsureConnection(ctx context.Context) error {
 	vc.finder = find.NewFinder(vc.vimDriver)
 	vc.pm = performance.NewManager(vc.vimDriver)
 	vc.vm = view.NewManager(vc.vimDriver)
+	vc.apiVersion = vc.getVsphereAPIVersion()
```
Comment (djaglowski) [inline receiver/vcenterreceiver/client.go:79]: If we drop support for 6.7 we don't need this check. Perhaps there is a reasonable way to fail gracefully if the server does not support the new metric?

Comment (BominRahmani) [reply receiver/vcenterreceiver/client.go:79]: I think we can just drop support for 6.7 in this case. The problem with this metric is it doesn't really fail, rather it returns 0 if the server doesn't support the metric, but 0 in of itself is a valid value.

Comment (jlg-io) [reply receiver/vcenterreceiver/client.go:79]: As 6.7 is "End of General Support" since October 15, 2022 (https://core.vmware.com/blog/reminder-vsphere-6567-end-general-support), it would make sense to drop the support here.

File: receiver/vcenterreceiver/README.md (hunk 7 above)
Comment (BominRahmani) [reply receiver/vcenterreceiver/README.md:60]: Yeah, it seems to be accurate since there is also an update from [Vmware](https://core.vmware.com/blog/reminder-vsphere-6567-end-general-support) saying the same thing.

File: receiver/vcenterreceiver/client.go (hunk 10 above)
Comment (BominRahmani) [reply receiver/vcenterreceiver/client.go]: @djaglowski To drop support for a version do we need to make a new pr / add warnings anywhere? or are we allowed to just remove it from the readme.

File: .chloggen/add_vcenter_vm_cpu_readiness_metric.yaml
Comment (StefanKurek) [inline .chloggen/add_vcenter_vm_cpu_readiness_metric.yaml:1]: not sure, but might want to add a new change log file with the `change_type: breaking` to explain that we're dropping support for 6.7 here officially. not completely sure if this is appropriate, because we're not "actually" doing anything. it seems correct to be marked in the changelog for the next release though...

File: receiver/vcenterreceiver/client.go (hunk 10 above)
Comment (djaglowski) [reply receiver/vcenterreceiver/client.go]: @BominRahmani, a separate PR would be ideal but I think it's ok to do here as long as there's a separate changelog entry for it. You can just add a second changelog yaml file and note it there.

**Review by djaglowski — APPROVED** [review]
